dask_expr/collection.py

```diff
@@ -58,7 +58,14 @@ def _wrap_expr_op(self, other, op=None):
     assert op is not None
     if isinstance(other, FrameBase):
         other = other.expr
-    return new_collection(getattr(self.expr, op)(other))
+    left = self.expr
+
+    if not isinstance(other, expr.Expr):
+        pass
+    else:
+        left, other = _maybe_align_partitions([left, other])
```

Comment:
Given my recent comment, I think that we maybe want some abstract `AlignDivisions` expression here or something.  We don't specify exactly what the divisions should be (because we genuinely don't know them yet and don't want to claim that we do) but we know that, when the time comes, we'll want to align the inputs along those divisions.

Maybe we keep the check if all of the divisions are the same though. This isn't quite correct (maybe one side's divisions could shift from the other's) but I wouldn't want to inject a ton of AlignDivisions expressions everywhere.  Thoughts?

Reply:
This ``AlignDivisions`` expression would set its ``_divisions`` to ``(None, None)`` then? It would only calculate its divisions when we construct the actual graph after all optimisations are through? Just making sure that I understand you correctly.

Couple of things come to mind:

- We potentially end up injecting this expression in a lot of places where it might not be necessary? e.g.:

```
x = a - b. # injects abstract ``AlignDivisions`` because they don't have the same structure 
y = x + c  # ``divisions`` of ``x``  are unknown, so inject another one?
```

- I am assuming that we would use the same pattern for ``align`` and friends where the alignment was specifically requested?
- I am wondering whether or not we shoot ourselves in the foot because follow up operations don't know the divisions of the previous object?

Reply:
I think that broadly we should try to use divisions as little as possible when building and operating on expressions.  Divisions are lazy for a reason.  They should mostly be used when constructing the graph.